### PR TITLE
ENG-690: `OakCookieSettingsModal`

### PR DESCRIPTION
**Pull request by carlmw**
## TODO

- [x] Replace `OakLink` in 3rd party policy links with `OakSecondaryLink` when this is merged https://github.com/oaknational/oak-components/pull/138

# How to review this PR

- Check [component hierarchy](https://miro.com/app/board/uXjVNnKBgyk=/?share_link_id=59445593794) is followed correctly
- Check the design [Heuristics](https://components.thenational.academy/?path=/docs/docs-howtodesigncomponents--docs#heuristics-for-component-design) have been followed
- Check [naming conventions](https://components.thenational.academy/?path=/docs/docs-namingconventions--docs) have been applied
- Check for these gotchyas:
  - Missing exports for Oak components
  - Accidental export of Internal components
  - Snapshots of unexpected components have been modified
  - Circular dependencies
  - Code duplication (via not using base components)
  - Non-functional storybook for this or related components
  - Sensitve files changed eg. atoms, or style tokens
  - Relative imports
  - Default exports

# Add your PR description below

Adds the cookie settings modal ready to be wired up to some state management. Includes form handling logic to capture the consent settings from the accordion.

https://github.com/oaknational/oak-components/assets/122096/af3cce53-2e24-43a3-92b9-1ba92d556d14

## Link to the design doc

https://www.figma.com/file/Ru5PP6cdyDTDdqmQpNwXtl/Cookie-Banner?type=design&node-id=1-11625&mode=dev

## A link to the component in the deployment preview

https://deploy-preview-139--lively-meringue-8ebd43.netlify.app/?path=/docs/components-organisms-oakcookiesettingsmodal--docs

## Testing instructions

There is no backend so particular focus should probably be made on how it behaves across the 3 breakpoints.

## ACs

NB i'm going to tackle the remaining AC in a follow PR that adds a little state management for both the modal and banner

 - [ ]  ~~Clicking “Allow all” checks all non-essential consent options~~
- [ ]  ~~Clicking “Reject non-essential cookies”~~
    - [ ]  ~~Unchecks non-essential cookies~~
    - [ ]  ~~Removes/does not set consent for non-essential cookies~~
    - [ ]  ~~Closes the consent modal~~
- [ ]  ~~Clicking “Confirm my choices”~~
    - [ ]  ~~Adds and removes consent according to the selection in the accordion~~
    - [ ]  ~~Closes the consent modal~~
- [x]  Exported from oak-components as an organism

### CODE REVIEW

File: src/components/molecules/OakModal/OakModal.test.tsx [hunk 1]
```diff
@@ -8,6 +8,9 @@ import { OakModal } from "./OakModal";
 import renderWithTheme from "@/test-helpers/renderWithTheme";
 import { oakDefaultTheme } from "@/styles";
 import { OakThemeProvider } from "@/components/atoms";
+import { installMockIntersectionObserver } from "@/test-helpers";
+
+installMockIntersectionObserver();
```
Comment (carlmw) [inline src/components/molecules/OakModal/OakModal.test.tsx:13]: I was getting fed up writing and installing these little mocks into JSDOM so I moved them into our helpers. These might be useful to export from the library at some point.

File: src/components/molecules/OakModal/OakModal.tsx [hunk 2]
```diff
@@ -58,6 +59,7 @@ const FadeOutBox = styled(OakBox)<{ $state: TransitionStatus }>`
 `;
 
 const SlideInFlex = styled(OakFlex)<{ $state: TransitionStatus }>`
+  max-width: calc(100vw - ${parseSpacing("inner-padding-l")});
```
Comment (carlmw) [inline src/components/molecules/OakModal/OakModal.tsx:62]: This ensures that the modal leaves a little space peeking through to the right when it slides out (matching the designs as much as is practical)

File: src/components/organisms/OakCookieSettingsModal/OakCookieSettingsModal.tsx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                  <OakCheckBox
+                    id={`cookies-settings-${policy.id}-checkbox`}
+                    name="consent"
+                    displayValue=""
+                    value={policy.id}
+                    defaultChecked={
+                      policy.strictlyNecessary ||
+                      initialConsents[policy.id] === "granted"
+                    }
+                    disabled={policy.strictlyNecessary}
```
Comment (carlmw) [inline src/components/organisms/OakCookieSettingsModal/OakCookieSettingsModal.tsx:177]: If a policy is strictly necessary it cannot be unchecked.

File: src/components/molecules/OakAccordion/OakAccordion.tsx [hunk 4]
```diff
@@ -69,6 +69,7 @@ export const OakAccordion = ({
       >
         <StyledOakFlex
           as="button"
+          type="button"
```
Comment (carlmw) [inline src/components/molecules/OakAccordion/OakAccordion.tsx:72]: Whoops 🤦🏻‍♂️

File: src/components/atoms/InternalCheckBox/InternalCheckBox.tsx [hunk 5]
```diff
@@ -41,6 +42,8 @@ export type BaseCheckBoxProps = {
   onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
   onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void;
   onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
+  "aria-label"?: string;
+  "aria-labelledby"?: string;
```
Comment (carlmw) [inline src/components/atoms/InternalCheckBox/InternalCheckBox.tsx:46]: Ideally the remaining input props should be being spread here. This isn't currently the case so i'm adding these to the white-list. I think we could make this component accept all the props of a vanilla checkbox going forward.

File: src/components/organisms/OakCookieSettingsModal/OakCookieSettingsModal.tsx [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  policies,
+  initialConsents,
+}: OakCookieSettingsModalProps) => {
+  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    const formData = new FormData(event.currentTarget);
+    const newConsents: Consents = {};
+
+    for (const policy of policies) {
+      newConsents[policy.id] = policy.strictlyNecessary ? "granted" : "denied";
```
Comment (carlmw) [inline src/components/organisms/OakCookieSettingsModal/OakCookieSettingsModal.tsx:98]: `strictlyNecessary` policies will always be set as `granted`

File: src/components/organisms/OakCookieSettingsModal/OakCookieSettingsModal.tsx [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    for (const policyId of formData.getAll("consent")) {
+      newConsents[policyId.toString()] = "granted";
```
Comment (carlmw) [inline src/components/organisms/OakCookieSettingsModal/OakCookieSettingsModal.tsx:102]: Pick all the checked policies from our form data and mark them as granted.

File: src/components/organisms/OakCookieSettingsModal/OakCookieSettingsModal.tsx [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  strictlyNecessary: boolean;
+  /**
+   * List of 3rd parties that the policy allows data to be shared with.
+   */
+  parties: Party[];
+};
+type ConsentState = "granted" | "denied";
+type Consents = {
+  [policyId: string]: ConsentState;
+};
```
Comment (carlmw) [inline src/components/organisms/OakCookieSettingsModal/OakCookieSettingsModal.tsx:46]: The intent here is that these align with https://github.com/oaknational/oak-consent-manager as much as is practical.

**Review by rightsaidjames — APPROVED** [review]

**Review by mantagen — APPROVED** [review]
Great looking code. And the modal seems to work very well ❤️

**Pull request merged**